compiler/rustc_mir_transform/src/const_prop.rs

```diff
@@ -379,24 +380,22 @@ impl<'mir, 'tcx> ConstPropagator<'mir, 'tcx> {
             ConstPropMachine::new(only_propagate_inside_block_locals, can_const_prop),
         );
 
-        let ret = ecx
+        let ret_layout = ecx
             .layout_of(body.return_ty().subst(tcx, substs))
             .ok()
-            // Don't bother allocating memory for ZST types which have no values
-            // or for large values.
-            .filter(|ret_layout| {
-                !ret_layout.is_zst() && ret_layout.size < Size::from_bytes(MAX_ALLOC_LIMIT)
-            })
-            .map(|ret_layout| {
-                ecx.allocate(ret_layout, MemoryKind::Stack)
-                    .expect("couldn't perform small allocation")
-                    .into()
-            });
+            // Don't bother allocating memory for large values.
+            .filter(|ret_layout| ret_layout.size < Size::from_bytes(MAX_ALLOC_LIMIT))
+            .unwrap_or_else(|| ecx.layout_of(tcx.types.unit).unwrap());
+
+        let ret = ecx
+            .allocate(ret_layout, MemoryKind::Stack)
+            .expect("couldn't perform small allocation")
+            .into();
```

Comment:
We definitely had a fair amount of issues by defaulting conceptually unused/unreachable places to a `()` in the past. So it warrants asking:

Is this actually valid? Maybe add a comment why doing so is fine (and/or documentation to `push_stack_frame`?

Reply:
I'm not so convinced it is valid actually, but I figure it can't be more wrong than what we were doing already - clearly something was already enforcing that we won't try and use this place if we're either not supposed to return or the size is too big. I don't really understand the contract here either though, maybe @RalfJung does? I'd be happy to add comments to `push_stack_frame` documenting under which conditions it's ok to put in nonsense for the return place

Reply:
Sorry, const prop is doing a lot of magic that I never really looked at. :/  In terms of just a MIR diff, what is this about?

Reply:
This isn't really an MIR diff. Before this PR, const prop had a rule that it would not try to do any evaluation for things that were greater than some maximum size. If the return destination for function call terminators violated this rule, it would instead pass `None` (which was normally reserved for cases where the function is guaranteed to diverge), and presumably ensure that this doesn't cause problems elsewhere. Because passing `None` is no longer an option, I am instead passing an allocation of `()`, and hoping that whatever was preventing bugs before will continue to do so now.

Reply:
This is very specific to const-prop implementation details then, which I am not familiar with. I just know that const-prop does keep causing problems since it uses the interpreter in funky ways...